### ISSUE TITLE
html/template: data race in Execute

### ISSUE THREAD

**Issue by bep** [golang/go#39807]
```bash
go version go1.14.3 darwin/amd64
```

Building the below program with `go build -race` and then running it fails:

```go
package main

import (
	"fmt"
	"html/template"
	"io/ioutil"
	"log"
	"sync"
)

func main() {
	var wg sync.WaitGroup

	jsTempl := `
{{- define "jstempl" -}}
var foo = "bar";
{{- end -}}
<script type="application/javascript">
{{ template "jstempl" $ }}
</script>
`

	tpl := template.New("")
	_, err := tpl.New("templ.html").Parse(jsTempl)
	if err != nil {
		log.Fatal(err)
	}

	const numTemplates = 20

	for i := 0; i < numTemplates; i++ {
		_, err = tpl.New(fmt.Sprintf("main%d.html", i)).Parse(`{{ template "templ.html" . }}`)
		if err != nil {
			log.Fatal(err)
		}
	}

	for i := 1; i <= 10; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < numTemplates; j++ {
				templ := tpl.Lookup(fmt.Sprintf("main%d.html", j))
				if err := templ.Execute(ioutil.Discard, nil); err != nil {
					log.Fatal(err)
				}

			}
		}()
	}

	wg.Wait()
}

```

<details>
<summary>WARNING: DATA RACE</summary>
<pre><code>
WARNING: DATA RACE
Write at 0x00c00011c540 by goroutine 32:
  runtime.mapassign_faststr()
      /Users/bep/dev/go/dump/go/src/runtime/map_faststr.go:202 +0x0
  text/template.(*Template).associate()
      /Users/bep/dev/go/dump/go/src/text/template/template.go:227 +0x1a8
  text/template.(*Template).AddParseTree()
      /Users/bep/dev/go/dump/go/src/text/template/template.go:133 +0x2cc
  html/template.(*escaper).commit()
      /Users/bep/dev/go/dump/go/src/html/template/escape.go:810 +0x2c5
  html/template.escapeTemplate()
      /Users/bep/dev/go/dump/go/src/html/template/escape.go:38 +0x324
  html/template.(*Template).escape()
      /Users/bep/dev/go/dump/go/src/html/template/template.go:102 +0x349
  html/template.(*Template).Execute()
      /Users/bep/dev/go/dump/go/src/html/template/template.go:119 +0x3c
  main.main.func1()
      /Users/bep/dev/go/bep/temp/main.go:46 +0x177

Previous read at 0x00c00011c540 by goroutine 28:
  runtime.mapaccess1_faststr()
      /Users/bep/dev/go/dump/go/src/runtime/map_faststr.go:12 +0x0
  text/template.(*state).walkTemplate()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:399 +0x13b
  text/template.(*state).walk()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:268 +0x2be
  text/template.(*state).walk()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:263 +0x195
  text/template.(*Template).execute()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:220 +0x2fb
  text/template.(*Template).Execute()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:203 +0xd2
  html/template.(*Template).Execute()
      /Users/bep/dev/go/dump/go/src/html/template/template.go:122 +0x7f
  main.main.func1()
      /Users/bep/dev/go/bep/temp/main.go:46 +0x177

Goroutine 32 (running) created at:
  main.main()
      /Users/bep/dev/go/bep/temp/main.go:42 +0x2de

Goroutine 28 (running) created at:
  main.main()
      /Users/bep/dev/go/bep/temp/main.go:42 +0x2de
==================
==================
WARNING: DATA RACE
Write at 0x00c0000ac0b0 by goroutine 32:
  text/template.(*Template).associate()
      /Users/bep/dev/go/dump/go/src/text/template/template.go:227 +0x1bd
  text/template.(*Template).AddParseTree()
      /Users/bep/dev/go/dump/go/src/text/template/template.go:133 +0x2cc
  html/template.(*escaper).commit()
      /Users/bep/dev/go/dump/go/src/html/template/escape.go:810 +0x2c5
  html/template.escapeTemplate()
      /Users/bep/dev/go/dump/go/src/html/template/escape.go:38 +0x324
  html/template.(*Template).escape()
      /Users/bep/dev/go/dump/go/src/html/template/template.go:102 +0x349
  html/template.(*Template).Execute()
      /Users/bep/dev/go/dump/go/src/html/template/template.go:119 +0x3c
  main.main.func1()
      /Users/bep/dev/go/bep/temp/main.go:46 +0x177

Previous read at 0x00c0000ac0b0 by goroutine 28:
  text/template.(*state).walkTemplate()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:399 +0x14e
  text/template.(*state).walk()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:268 +0x2be
  text/template.(*state).walk()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:263 +0x195
  text/template.(*state).walkTemplate()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:413 +0x3cd
  text/template.(*state).walk()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:268 +0x2be
  text/template.(*state).walk()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:263 +0x195
  text/template.(*Template).execute()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:220 +0x2fb
  text/template.(*Template).Execute()
      /Users/bep/dev/go/dump/go/src/text/template/exec.go:203 +0xd2
  html/template.(*Template).Execute()
      /Users/bep/dev/go/dump/go/src/html/template/template.go:122 +0x7f
  main.main.func1()
      /Users/bep/dev/go/bep/temp/main.go:46 +0x177

Goroutine 32 (running) created at:
  main.main()
      /Users/bep/dev/go/bep/temp/main.go:42 +0x2de

Goroutine 28 (running) created at:
  main.main()
      /Users/bep/dev/go/bep/temp/main.go:42 +0x2de
==================
Found 2 data race(s)
</pre></code>
</detail>

**Comment by davecheney** [comment]
@bep thanks for the issue. Are you able to check Go 1.13 and see if it is affected? (This will determine if this is a blocker for 1.15 or not) 

/cc @dmitshur

**Comment by bep** [comment]
It also fails on go1.13.12.

**Comment by empijei** [comment]
Some additional info:
* This does not affect "text/template"
* This also affects `ExecuteTemplate` so it is not just an issue with the lookup+execute calls.
* This does not happen if the templates are executed at least once before the concurrent execution. This is probably due to a lazy computation since the race happens in a `escapeTemplate` call, which is only executed once per template.
* My best guess is that the computation is lazy because it involves calling nested templates but I'd need to take a better look at this. The escape should be happening while holding a lock or should be executed at parse time if possible.

As a quick workaround running the templates once before they are run concurrently addresses the issue (see code below), but we need to properly fix this.

Thanks for reporting.

```go
package main

import (
	"fmt"
	"html/template"
	"io/ioutil"
	"log"
	"sync"
)

func main() {
	var wg sync.WaitGroup

	jsTempl := `
{{- define "jstempl" -}}
var foo = "bar";
{{- end -}}
<script type="application/javascript">
{{ template "jstempl" $ }}
</script>
`

	tpl := template.New("")
	_, err := tpl.New("templ.html").Parse(jsTempl)
	if err != nil {
		log.Fatal(err)
	}

	const numTemplates = 20

	for i := 0; i < numTemplates; i++ {
		_, err = tpl.New(fmt.Sprintf("main%d.html", i)).Parse(`{{ template "templ.html" . }}`)
		if err != nil {
			log.Fatal(err)
		}
	}

        // Temporary workaround to prevent the race.
	for j := 0; j < numTemplates; j++ {
		if err := tpl.ExecuteTemplate(ioutil.Discard, fmt.Sprintf("main%d.html", j), nil); err != nil {
			log.Fatal(err)
		}
	}

	for i := 1; i <= 10; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < numTemplates; j++ {
				if err := tpl.ExecuteTemplate(ioutil.Discard, fmt.Sprintf("main%d.html", j), nil); err != nil {
					log.Fatal(err)
				}
			}
		}()
	}

	wg.Wait()
}
```

**Comment by andreasf** [comment]
This appears to be the same data race causing gohugoio/hugo#7293, where it is triggered from text/template.

Hugo have a temporary fork of text/template included in their codebase. I attempted to fix the data race in the fork: gohugoio/hugo#7507. They'd rather have it upstreamed though.

**Comment by andreasf** [comment]
I pulled the patch out of Hugo's fork of text/template and changed it a bit:

* Separate lock for `tmpl` instead of reusing `muFunc`—not sure if that's best
* Use lock everywhere `tmpl` is accessed

Personally, I'd prefer a single lock as it's simpler and safer. What do you think, should I file a PR?

**Comment by empijei** [comment]
Do you have a proposal to fix the race described in this bug report? If so please do send a CL my way :smile:

**Comment by rsc** [comment]
It looks to me like html/template's nameSpace.mu needs to be an RWLock and must be RLock'ed during t.text.Execute.
t.escape will need to RLock to decide if escaping is necessary and then RUnlock/Lock if so.

**Comment by bep** [comment]
I guess this will not get fixed in Go 1.16?

**Comment by ianlancetaylor** [comment]
The problem in text/template seems different to me.  That seems to require calling `AddParseTree` concurrently, but that method is not documented as supporting concurrent calls.  Maybe we should change that, but it doesn't seem related to this problem.

**Comment by bcmills** [comment]
See previously #2439, which was apparently the same problem but regressed due to the lack of a regression test (which is understandable given that the race detector debuted over a year later).

**Comment by bep** [comment]
@bcmills sure, but you (and with _you_ I don't mean specifically you) cannot at the sam time say that _data races are always a serious bug which need fixing_ and then brush this on to a future release when there is a working fix in place.

This whole situation means that I need to somehow maintain a fix for this in a fork, which is not something I do with joy and certainly not something I get paid to do.

/cc @spf13

**Comment by ianlancetaylor** [comment]
@bep 

> you (and with you I don't mean specifically you) cannot at the sam time say that data races are always a serious bug which need fixing and then brush this on to a future release when there is a working fix in place.

I understand that the length of time this has been open is frustrating, but I don't see anybody saying this.

Are you able to check whether https://golang.org/cl/274450 fixes the problems that you are seeing?

**Comment by ianlancetaylor** [comment]
I think we're going to have to roll back the fix for this.  I don't see a way to avoid #43855.

**Comment by ianlancetaylor** [comment]
The fix was rolled back, so reopening this issue.

**Comment by ianlancetaylor** [comment]
The basic issue here is that html/template expects to be able to call the text/template `Template` methods `Execute` and `AddParseTree` concurrently.  We can in principle fix this by adding locking to html/template, but it's hard because user-defined template functions, including calls to data type methods, can themselves call `Execute` and, therefore, `AddParseTree`.

My attempt to fix this, following the suggestion at https://github.com/golang/go/issues/39807#issuecomment-703966128, failed because nothing released the read lock held during a method call, so if the method call called `Execute`, it led to a deadlock.

Given html/template's needs, I think we kind of do have to fix this in text/template, along the lines of https://golang.org/cl/257817.  It's certainly an abstraction violation to change text/template to fix an html/template problem.  But given that html/template needs to concurrently call `Execute` and `AddParseTree`, I don't yet see a different workable approach.

**Comment by bcmills** [comment]
Following my comment in https://golang.org/cl/c/go/+/274450/4/src/html/template/template.go#138, I think it is possible to fix this race by escaping all of the templates at the first call to `Execute`, guarded by a `sync.Once`. But an attempt to do that apparently causes [`TestAddParseTreeHTML`](https://github.com/golang/go/blob/1b09d430678d4a6f73b2443463d11f75851aba8a/src/html/template/clone_test.go#L18-L33) to fail.

I suspect that the failure observed in `TestAddParseTreeHTML` could be addressed by escaping the templates in topological (instead of arbitrary) order.

However, I suspect that that failure indicates a fundamental inconsistency in the API: if the template `b` cannot be escaped outside the context of its invocation within template `a`, doesn't that imply that the same template might need to be escaped _multiple_ times, in different ways depending on the context? But the map data structures in the `html/template` package do not appear to include that context as part of the map key — so either the failure to escape the template in the test is spurious, or a different test ought to reveal a condition in which the template is mis-escaped.

**Comment by ianlancetaylor** [comment]
I don't think there is any inconsistency in the API.  When a top-level template is escaped, all the `{{template NAME}}` invocations are rewritten to invoke a template escaped in the proper context.  That escaping is indeed context-dependent, the escaped template is given a context-dependent name (see the `mangle` method in context.go), and the escaped template is not marked in a way that is seen if the template is invoked at the top level or in some other context.

A topological sort might work, if we don't care about the extra work.

**Comment by bep** [comment]
@ianlancetaylor if a building is about to tip over, you put some support pillars in place. I reviewed a perfectly fine CR (https://golang.org/cl/257817) that would have made a great pillar until you could get a proper foundation in place. **Would you have released Go 1.16 with a similar bug in the net/http package**?

**Comment by ianlancetaylor** [comment]
If the net/http bug had existed for several releases, and if it were this late in the release cycle, then, yes, I would not have blocked Go 1.16 from being released with such a bug.

**Comment by bep** [comment]
@ianlancetaylor https://github.com/golang/go/issues/43822#issuecomment-768639142

**Comment by ianlancetaylor** [comment]
@bep I'm sorry you are upset about this issue.  I understand that it is frustrating.  It's frustrating to me too.  I've probably put 40 hours into fixing this issue, and so far I have failed.  That's a lot of time for me.

That said, the comment you mention is different.  I'm not blocking 1.16 because of that bug.  And, to use your metaphor, that CL is not a case of putting in support pillars until there is a proper foundation.  That CL is, I believe, the correct fix.

**Comment by ianlancetaylor** [comment]
OK, I think this is finally fixed, and the fix will be in the 1.17 release.  I hope.  Sorry for the lengthy work on this.
